drivers/hmis/app/graphql/types/forms/form_definition.rb

```diff
@@ -48,6 +51,54 @@ def definition
       )
     end
 
+    def project_matches
+      # Returns one match per project that this form applies to. (See InstanceProjectMatch for match ranking logic.)
+      # This DOES still return a match for a project if the match is overridden by a more specific rule on another form.
+      # For example, if:
+      # - Intake form A has a rule that specifies that it's used for all projects
+      # - Intake form B has a rule that specifies that it's used for all Emergency Shelter projects
+      # - Intake form C has a rule that specifies that it's used for Project X, an ES project
+      # ...then forms A and B would still return a match for Project X.
+
+      instance_scope = object.instances.active
+
+      # Fetching all projects is slow, so do it only once and pass the scope to project_matches below.
+      project_scope = Hmis::Hud::Project.hmis.preload(:organization)
+
+      # Preload funders only if any of these rules are funder-based (micro-optimizing)
+      project_scope = project_scope.preload(:funders) if instance_scope.where.not(funder: nil).exists?
```

Comment:
micro optimizing the micro-optimization: `instance_scope.find { |inst| inst.funder.present? }`

We hit the database once to get all the instances (line 63). We expect the list of instances to be relatively small, so we can operate on it in-memory. (As opposed to hitting the database again to ask if any of them have a funder, which is what the `exists?` query will do).

There are plenty of places below where we choose whether to hit the db or use memory:
`instance_scope.defaults.exists?` <= db
`instance_scope.map(&:project_type)` <= memory (db would be to use `pluck`)
`instance_scope.select ... ` <= memory (db would be `where`)


I hope I didn't over-explain that... just taking the opportunity to point out something it can be helpful to think about when writing code in the graphql resolvers. None of these examples here seem that important to me -- but it can be problematic if we are really hammering the API for something that we could easily do in-memory, for example.

Reply:
Yes this makes sense, noo worries about over-explaining, the level of explanation is perfect! ☺️ 

I've made the update you suggested here. I'm assuming that it's not worth it to replace the `instance_scope.defaults.exists?` db call with an equivalent in-memory matcher. But would be happy to do so if you disagree.